File: hw2/grades/lib/grades/calculator.ex
defmodule Grades.Calculator do
  
  def percentage_grade(%{homework: homework, labs: labs, midterm: midterm, final: final}) do
    avg_homework = avg(%{list: homework})
    avg_labs = avg(%{list: labs})

	mark = calculate_grade(%{lab: avg_labs, hw: avg_homework, mdt: midterm, fin: final})
    round(mark * 100)
  end

  def letter_grade(%{homework: homework, labs: labs, midterm: midterm, final: final}) do
    avg_homework = avg(%{list: homework})
    avg_labs = avg(%{list: labs})

    avg_exams = (midterm + final) / 2

    num_labs =
      labs
      |> Enum.reject(fn mark -> mark < 0.25 end)
      |> Enum.count()

    if failed_to_participate?(%{hw: avg_homework, ex: avg_exams, nlab: num_labs}) do
      "EIN"
    else
	  mark = calculate_grade(%{lab: avg_labs, hw: avg_homework, mdt: midterm, fin: final})

      cond do
        mark > 0.895 -> "A+"
        mark > 0.845 -> "A"
        mark > 0.795 -> "A-"
        mark > 0.745 -> "B+"
        mark > 0.695 -> "B"
        mark > 0.645 -> "C+"
        mark > 0.595 -> "C"
        mark > 0.545 -> "D+"
        mark > 0.495 -> "D"
        mark > 0.395 -> "E"
        :else -> "F"
      end
    end
  end

  def numeric_grade(%{homework: homework, labs: labs, midterm: midterm, final: final}) do
    avg_homework = avg(%{list: homework})
    avg_labs = avg(%{list: labs})

    avg_exams = (midterm + final) / 2

    num_labs =
      labs
      |> Enum.reject(fn mark -> mark < 0.25 end)
      |> Enum.count()

    if failed_to_participate?(%{hw: avg_homework, ex: avg_exams, nlab: num_labs}) do
      0
    else
      mark = calculate_grade(%{lab: avg_labs, hw: avg_homework, mdt: midterm, fin: final})

      cond do
        mark > 0.895 -> 10
        mark > 0.845 -> 9
        mark > 0.795 -> 8
        mark > 0.745 -> 7
        mark > 0.695 -> 6
        mark > 0.645 -> 5
        mark > 0.595 -> 4
        mark > 0.545 -> 3
        mark > 0.495 -> 2
        mark > 0.395 -> 1
        :else -> 0
      end
    end
  end
  
  def avg(%{list: list}) do
	if Enum.count(list) == 0 do
		0
	else
		Enum.sum(list) / Enum.count(list)
	end
  end
  
  def failed_to_participate?(%{hw: hw, ex: ex, nlab: nlab}) do
	if hw < 0.4 || ex < 0.4 || nlab < 3 do
		true
	else
		false
	end
  end
  
  def calculate_grade(%{lab: lab, hw: hw, mdt: mdt, fin: fin}) do
	0.2 * lab + 0.3 * hw + 0.2 * mdt + 0.3 * fin
  end
end

File: hw2-midnight/grades_java (optional)/src/Calculator.java
public class Calculator {	
	
	//Note: as "final" is a reserved keyword in Java, the value was renamed finterm. 
	public static int percentageGrade(float[] homework, float[] labs, float midterm, float finterm) {
		float avgHomework = (homework.length == 0) ? 0 : (arraySum(homework) / homework.length); 
		float avgLabs = (labs.length == 0) ? 0 : (arraySum(labs) / labs.length); 
		
		return (int) Math.round((100)*(0.2 * avgLabs + 0.3 * avgHomework + 0.2 * midterm + 0.3 * finterm));
	}
	
	public static String letterGrade(float[] homework, float[] labs, float midterm, float finterm) {
		float avgHomework = (homework.length == 0) ? 0 : (arraySum(homework) / homework.length); 
		float avgLabs = (labs.length == 0) ? 0 : (arraySum(labs) / labs.length); 
		float avgExams = (midterm + finterm) / 2;
		
		int numLabs = 0;
		for(float mark : labs) if(mark >= 0.25) numLabs++;
		
		if(avgHomework < 0.4 || avgExams < 0.4 || numLabs < 3) return "EIN";
		else {
			float mark = 0.2f * avgLabs + 0.3f * avgHomework + 0.2f * midterm + 0.3f * finterm;
			
			//Unfortunately, cannot use a switch statement for floats...
			if(mark > 0.895) return "A+";
			if(mark > 0.845) return "A";
			if(mark > 0.795) return "A-";
			if(mark > 0.745) return "B+";
			if(mark > 0.695) return "B";
			if(mark > 0.645) return "C+";
			if(mark > 0.595) return "C";
			if(mark > 0.545) return "D+";
			if(mark > 0.495) return "D";
			if(mark > 0.395) return "E";
			return "F";
		}
	}
	
	public static int numericGrade(float[] homework, float[] labs, float midterm, float finterm) {
		float avgHomework = (homework.length == 0) ? 0 : (arraySum(homework) / homework.length); 
		float avgLabs = (labs.length == 0) ? 0 : (arraySum(labs) / labs.length); 
		float avgExams = (midterm + finterm) / 2;
		
		int numLabs = 0;
		for(float mark : labs) if(mark >= 0.25) numLabs++;
		
		if(avgHomework < 0.4 || avgExams < 0.4 || numLabs < 3) return 0;
		else {
			float mark = 0.2f * avgLabs + 0.3f * avgHomework + 0.2f * midterm + 0.3f * finterm;
			
			//Unfortunately, cannot use a switch statement for floats...
			if(mark > 0.895) return 10;
			if(mark > 0.845) return 9;
			if(mark > 0.795) return 8;
			if(mark > 0.745) return 7;
			if(mark > 0.695) return 6;
			if(mark > 0.645) return 5;
			if(mark > 0.595) return 4;
			if(mark > 0.545) return 3;
			if(mark > 0.495) return 2;
			if(mark > 0.395) return 1;
			return 0;
		}
	}
	
	//Replacement for Elexir's "Enum.sum" method
	private static float arraySum(float[] arr) {
		float sum = 0;
		for(float n : arr) sum += n;
		return sum;
	}
}
